instrumentation/apache-pulsar/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pulsar/ConsumerImplInstrumentation.java

```diff
@@ -0,0 +1,131 @@
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import io.opentelemetry.javaagent.instrumentation.pulsar.info.ClientEnhanceInfo;
+import io.opentelemetry.javaagent.instrumentation.pulsar.info.MessageEnhanceInfo;
+import io.opentelemetry.javaagent.instrumentation.pulsar.textmap.MessageTextMapGetter;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.impl.ConsumerImpl;
+import org.apache.pulsar.client.impl.MessageImpl;
+import org.apache.pulsar.client.impl.PulsarClientImpl;
+
+public class ConsumerImplInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("org.apache.pulsar.client.impl.ConsumerImpl");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    String klassName = ConsumerImplInstrumentation.class.getName();
+
+    transformer.applyAdviceToMethod(isConstructor(), klassName + "$ConsumerImplConstructorAdviser");
+
+    transformer.applyAdviceToMethod(
+        isMethod()
+            .and(isProtected())
+            .and(named("messageProcessed"))
+            .and(takesArgument(0, named("org.apache.pulsar.client.api.Message"))),
+        klassName + "$ConsumerImplMethodAdviser");
+  }
+
+  @SuppressWarnings("unused")
+  public static class ConsumerImplConstructorAdviser {
+
+    @Advice.OnMethodExit
+    public static void before(
+        @Advice.This ConsumerImpl<?> consumer,
+        @Advice.Argument(value = 0) PulsarClient client,
+        @Advice.Argument(value = 1) String topic) {
+
+      PulsarClientImpl pulsarClient = (PulsarClientImpl) client;
+      String url = pulsarClient.getLookup().getServiceUrl();
+
+      ClientEnhanceInfo info = new ClientEnhanceInfo(topic, url);
+      ClientEnhanceInfo.virtualField(consumer, info);
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class ConsumerImplMethodAdviser {
+
+    @Advice.OnMethodEnter
+    public static Scope before(
+        @Advice.This ConsumerImpl<?> consumer, @Advice.Argument(value = 0) Message<?> message) {
+      ClientEnhanceInfo info = ClientEnhanceInfo.virtualField(consumer);
+      if (null == info) {
+        return Scope.noop();
+      }
+
+      MessageImpl<?> messageImpl = (MessageImpl<?>) message;
+      Context context =
+          PROPAGATOR.extract(Context.current(), messageImpl, MessageTextMapGetter.INSTANCE);
+
+      return TRACER
+          .spanBuilder("Pulsar://ConsumerImpl/messageProcessed")
```

Comment:
You really should be using Instrumenter api. In this repository we have converted all instrumentations to it. Just using it will help you get many of the small details like span naming ([opentelemetry messaging semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) states that messaging spans should be named `<destination name> <operation name>`) right without extra effort. Search for usages of `MessagingAttributesExtractor` using the instrumenter api is really quite straightforward.